assets/deepen_shallow_clone_until_ref_is_found_then_check_out

```diff
@@ -0,0 +1,50 @@
+#!/bin/bash
+# vim: set ft=sh
+
+set -e
+
+readonly max_depth=128
+
+declare depth="$1"
+readonly ref="$2"
+readonly tagflag="$3"
+
+# A shallow clone may not contain the Git commit $ref:
+# 1. The depth of the shallow clone is measured backwards from the latest
+#    commit on the given head (master or branch), and in the meantime there may
+#    have been more than $depth commits pushed on top of our $ref.
+# 2. If there's a path filter (`paths`/`ignore_paths`), then there may be more
+#    than $depth such commits pushed to the head (master or branch) on top of
+#    $ref that are not affecting the filtered paths.
+#
+# In either case we try to deepen the shallow clone until we find $ref, reach
+# the max depth of the repo, or give up after a given depth and resort to deep
+# clone.
+
+git_dir="$(git rev-parse --git-dir)"
+readonly git_dir
+
+while ! git checkout -q "$ref" &>/dev/null; do
```

Comment:
If it were me, I would change this `checkout` to be one of the different commands for checking if a commit exists.  Now that the checkout happens at the end, you don't need it here, and it's mildly clearer for an if test not to have side effects.

I prefer `git cat-file -e "$ref"` because it's branch independent (and -e means "exit 0 if valid, exit 1 if not"), but you can also use `git rev-list "HEAD..$ref"` or I've actually seen people advocate for `git merge-base "$ref" "$ref"`, which I find kind of goofy.